crates/ide_completion/src/completions/record.rs

```diff
@@ -14,12 +14,31 @@ pub(crate) fn complete_record(acc: &mut Completions, ctx: &CompletionContext) ->
             | ImmediateLocation::RecordExprUpdate(record_expr),
         ) => {
             let ty = ctx.sema.type_of_expr(&Expr::RecordExpr(record_expr.clone()));
+
             let default_trait = ctx.famous_defs().core_default_Default();
-            let impl_default_trait = default_trait.zip(ty).map_or(false, |(default_trait, ty)| {
-                ty.original.impls_trait(ctx.db, default_trait, &[])
-            });
+            let impl_default_trait =
+                default_trait.zip(ty.as_ref()).map_or(false, |(default_trait, ty)| {
+                    ty.original.impls_trait(ctx.db, default_trait, &[])
+                });
 
-            let missing_fields = ctx.sema.record_literal_missing_fields(record_expr);
+            let missing_fields = match ty.and_then(|t| t.adjusted().as_adt()) {
+                Some(hir::Adt::Union(un)) => {
+                    // ctx.sema.record_literal_missing_fields will always return
+                    // an empty Vec on a union literal. This is normally
+                    // reasonable, but here we'd like to present the full list
+                    // of fields if the literal is empty.
+                    let were_fields_specified = record_expr
+                        .record_expr_field_list()
+                        .and_then(|fl| fl.fields().next())
+                        .is_some();
+
+                    match were_fields_specified {
+                        false => un.fields(ctx.db).into_iter().map(|f| (f, f.ty(ctx.db))).collect(),
+                        true => vec![],
+                    }
+                }
+                _ => ctx.sema.record_literal_missing_fields(record_expr),
+            };
             if impl_default_trait && !missing_fields.is_empty() && ctx.path_qual().is_none() {
```

Comment:
We should prevent this branch from firing for unions as well as the following, the `..Base` expression make no sense on unions, they aren't valid at all

Reply:
Might be better to just split the whole block into "is it a union" and "is it a struct" branches at that point...

Reply:
Might be the case ye

Reply:
alright, there you go